### PR TITLE
[#126] Implement `total_supply` off-chain view

### LINKED ISSUE #126: Implement `total_supply` off-chain view
# Clarification and motivation

TZIP-12 spec says that `total_supply` SHOULD be provided, so it may be useful to implement it. The tricky part is that we do not track total supply yet.

Note that it is likely that along with off-chain view, an on-chain view will have to be implemented as well because at the moment of writing we don't have a neat way to test an off-chain view against a contract (though such way may be added very soon, track morley-metadata), and also because there is a [suggestion](https://gitlab.com/tzip/tzip/-/issues/52) in `tzip` to use both types of views anyway
 
# Acceptance criteria

* `total_supply` off-chain view is implemented according to TZIP-12
* The new `totalSupply` field in the contract, if added, is covered with appropriate tests.

### PR DESCRIPTION
**Pull request by rinn7e**
## Description

Problem: TZIP-12 specs say that `total_supply` SHOULD be provided but we
do not track total supply yet.

Solution: Track total supply for all the tokens, add off-chain and
on-chain `total_supply` view, and add tests to cover all the cases.

## Related issue(s)

Resolves #126 

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

[//]: # (Add more docs here if you have them in the repository)
- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)
    - Haddock

#### Stylistic guide (mandatory)

- [x] My commits comply with [the following policy](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

### CODE REVIEW

**Comment by rinn7e** [comment]
@pasqu4le I've added the `get_total_supply` off-chain view now, I have to add on-chain view as well to be able to test this. This on-chain view will be removed later when it is possible to test off-chain view against a contract.

The MR is pretty much ready for review, only a few `TODO` related to documentation are left. Hope to get your review on this.

**Comment by pasqu4le** [comment]
I think we shouldn't necessarily remove the on-chain view for now.

The issue points at a [suggestion](https://gitlab.com/tzip/tzip/-/issues/52) that is still open, but aside from it if it doesn't create problems we can keep it for the extent of this issue and decide later if we want to remove it.

**Comment by rinn7e** [comment]
> I think we shouldn't necessarily remove the on-chain view for now.
> 
> The issue points at a [suggestion](https://gitlab.com/tzip/tzip/-/issues/52) that is still open, but aside from it if it doesn't create problems we can keep it for the extent of this issue and decide later if we want to remove it.

I've removed the comment mentioning about removing the on-chain view. 

I also handle various TODO and add proper doc.

**Comment by rinn7e** [comment]
I'll wait for #144 to be merged first before continue working on this to avoid merged conflict.

**Comment by rinn7e** [comment]
@pasqu4le I've rebase with the latest, and also fixing some issues that were not working before. This should be ready for review.

File: ligo/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs
Comment (pasqu4le) [inline ligo/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs:38]: Why the addition of this address balance(s) here?

I'd imagine that in a `defaultStorage` the ledger should be empty (as it was before)

File: ligo/src/permit.mligo
Comment (pasqu4le) [inline ligo/src/permit.mligo:63]:
```suggestion
  match Big_map.find_opt param.token_id store.total_supply with
    | None -> (failwith("FA2_TOKEN_UNDEFINED") : return)
    | Some v ->
      ( Tezos.transaction result 0mutez param.callback
        :: ([] : operation list)
      , store
      )
```
I'm surprised that the `:: ([] :operation list)` works BTW, shouldn't `Tezos.transaction` be wrapped in `[]`?

File: ligo/src/proposal.mligo
Comment (pasqu4le) [inline ligo/src/proposal.mligo:269]: Let's split this in more lines

Comment (pasqu4le) [inline ligo/src/proposal.mligo:269]: Let's split this in more lines

Comment (pasqu4le) [inline ligo/src/proposal.mligo:269]: Let's split this in more lines

File: ligo/src/token/fa2.mligo
Comment (pasqu4le) [inline ligo/src/token/fa2.mligo:42]: I think this should be splitted in steps, like:
- first `current_total_supply = ...` (put 0 in case of `None`)
- then `new_total_supply = ...` (fail on `None`)
- etc.

instead of indenting it so much.

File: src/BaseDAO/ShareTest/Common.hs
Comment (pasqu4le) [inline src/BaseDAO/ShareTest/Common.hs:57]: Nitpick, but `get` seems to indicate some fetching, I suggest renaming this just to `totalSupplyFromLedger`

File: ligo/src/startup.mligo
Comment (pasqu4le) [inline ligo/src/startup.mligo:103]: This should be mentioned as one of the entrypoints that doesn't directly traslate in the "module" comment on top.

**Review by pasqu4le — COMMENTED** [review]
A few suggestions, also the specifications docs need to be updated as well

File: ligo/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs
Comment (rinn7e) [reply ligo/haskell/test/Test/Ligo/BaseDAO/OffChainViews.hs:38]: This `defaultStorage` is used to test FA2 off-chain views:

```
  , testGroup "get_total_supply" $
    let checkTotalSupply st param =
          runView @Natural (DAO.getTotalSupplyView mc) st (ViewParam param)
    in
    [ testCase "Get unfrozen token total supply" $
        checkTotalSupply storage DAO.unfrozenTokenId
          @?= Right 100
    , testCase "Get frozen token total supply" $
        checkTotalSupply storage DAO.frozenTokenId
          @?= Right 200
    ]
```

Maybe renaming this to something like offchainViewStorage would be better.

File: ligo/src/permit.mligo
Comment (rinn7e) [reply ligo/src/permit.mligo:63]: Actually `::` here is a con, not type specifying. Could be quite confusing with the haskell syntax.

File: ligo/src/startup.mligo
Comment (rinn7e) [reply ligo/src/startup.mligo:103]: I've updated this, as well as other changes, this should be good now.

File: ligo/src/permit.mligo
Comment (pasqu4le) [reply ligo/src/permit.mligo:63]: Oh yeah :sweat_smile: syntax confusion

File: ligo/src/proposal.mligo
Comment (pasqu4le) [reply ligo/src/proposal.mligo]: I have no idea why this comment repeated 3 times, sorry about that

**Review by pasqu4le — COMMENTED** [review]
LGTM now, except for the specification that still need to be updated.

**Comment by rinn7e** [comment]
> LGTM now, except for the specification that still need to be updated.

I've updated both lorentz and ligo specification now. Please take a look, Pasquale.

**Review by pasqu4le — APPROVED** [review]

**Pull request merged**
